draft-ietf-cdni-ci-triggers-rfc8007bis.xml

```diff
@@ -876,86 +881,112 @@
+   If dCDN does remove triggers in a terminal state automatically, it MUST report the expiry timeout period, 
+   using an attribute "staleresourcetime" of the trigger collection resource (see 
+   <xref target="ci-trigger-collection-resource"/> for details.
    </t>
+
    <t>
-           It is RECOMMENDED that dCDN sets its expiry timeout to at least 24 hours. 
-        It is further RECOMMENDED that uCDN sets its trigger polling period to less than the expiry period, so it doesn't miss
-        trigger status updates before the "complete" or "failed" triggers are expired by dCDN.
+   It is RECOMMENDED that dCDN sets its staleresourcetime to at least 24 hours. 
+   It is further RECOMMENDED that uCDN sets its trigger polling period to less than this period, 
+   so it doesn't miss trigger status updates before the "complete" or "failed" triggers are expired by dCDN.
    </t>
+
    </section>
 
    <section title="Loop Detection and Prevention" anchor="loop-detection-and-prevention">
    <t>
-           Given three CDNs, A, B, and C, if CDNs B and C delegate delivery of CDN A's content to each other, 
-           CDN A's trigger creation requests could be passed between CDNs B and C in a loop. 
-           More complex networks of CDNs could contain similar loops involving more hops.
+   Given three CDNs, A, B, and C, if CDNs B and C delegate delivery of CDN A's content to each other, 
+   CDN A's trigger creation requests could be passed between CDNs B and C in a loop. 
+   More complex networks of CDNs could contain similar loops involving more hops.
    </t>
 
    <t>
-   To prevent and detect such CI/T loops, each CDN uses a <xref target="cdn-pid">CDN Provider ID (PID)</xref> to uniquely identify itself. 
-   In every trigger resource it creates or cascades, each CDN MUST append an array element
-   containing its CDN PID to a JSON array under an entry named "cdn-path". When receiving a trigger creation request, dCDN MUST check the
-   cdn-path and reject any trigger that already contains its own CDN PID in the cdn-path. Transit CDNs MUST check the "cdn-path" and
-   not cascade the trigger creation request to dCDNs that are already listed in the cdn-path.
+   When such CDN topologies become possible, it is RECOMMENDED that CDNs participating in it utilize 
+   <xref target="cdn-pid">CDN Provider ID (PID)</xref> to detect and prevent loops as follows:
+   <list style="symbols">
+   <t>
+   uCDNs that originate a new trigger request, SHOULD specify their CDN provider ID using 
+   trigger "cdn-path" attribute (see <xref target="ci-trigger-resource"/> for details).
+   </t>
+   <t>
+   dCDN that receives a trigger creation request that contains "cdn-path" attribute MUST reject a 
+   trigger that already contains its own CDN PID in the CDN path.
+   </t>
+   <t>
+   dCDN that cascades trigger requests to additional dCDNs (so-called "transit CDN"), SHOULD reject
+   triggers that have CDN PID of its downstream CDNs in their CDN path already.
```

Comment:
Is reject the correct behavior here?  It seems to me that if the dCDN has not seen the trigger before, it should process it rather than reject it.  Wouldn't it be more correct (as an improvement in efficiency) for it to not propagate the trigger to downstream CDNs in the CDN path already?

Reply:
we have two loop checks in the spec now - dCDN receives a trigger and sees itself in the path and this one - which means do not pass it to dCDNs that have already seen it
in my mind, if CDNs end up in a loop, this is "drop everything and fix the loop" situation, so i'd rather err on the side of failing triggers versus trying to salvage parts of processing

Reply:
In thinking about this, I was thinking about a possibility where 3 CDNs have relationships with one another, rather than just the tCDN having relationships with uCDN and dCDN.  In such a case, they could switch roles depending on the scenario, leading to a situation where a provider configured as a dCDN of a tCDN in one scenario could actually be a uCDN of an upstream tCDN in another scenario.  In such a case, we would not want to fail a trigger if the dCDN had not seen it before, even if it has a relationship with another dCDN.

Do I think this is a realistic scenario?  Not really at all.  I can't think of a valid use case for this.  But then again, I see the tCDN situation in general as being pretty atypical, and the diamond configuration being quite farfetched.

So, I don't feel strongly about this at all.  The solution you proposed would definitely put loop safety above more thorough processing of triggers, which is probably fine.

Reply:
Yes, I was thinking of the same scenario, this is the original thought framework of CDNi, where telecom-based CDNs federate with each other
I came up with one - really far-fetched - "diamond" scenario where a public CDN A delegates to open caching CDN B, which acts as tCDN and delegates it down to another open caching CDN C, which uses public CDN A for overflow capacity 
in this case the best outcome would be for CDN A to report successful trigger processing back to C back to B back to A, not to fail the trigger. 
so maybe uCDN shouldn't fail a trigger on behalf of its dCDN?

Reply:
Like you said, this is sounding *really* farfetched.

I'm not sure if the federated model by itself ever made sense, as I would think that major telecoms lack the incentive to make it happen.  Perhaps there are some more rare cases where a smaller telecom is directly partnered and peered with a larger telecom that it leases resources from, where there is a possibility for that to happen.  But most telecoms lack the technical knowledge to make this happen, leaving an opportunity for service providers (like my company) to facilitate the federation of multiple service provider networks.  But even in this case, it is rare that service providers are actually sharing resources with one another, and our network mostly appears like a single OC network to other CDNs, not a federation of different OC networks.

After giving your example some thought, yes, maybe this could happen in really rare cases.  Perhaps OC CDN C is also a content provider that offers its own cable package over streaming, which it makes available via its own OC network.  And, perhaps it partners with public CDN A for off-network access, and for specific cases where it has insufficient OC capacity in its own network.  And, perhaps some of those off-network subscribers are temporarily connected through the provider of OC CDN B, so either OC CDN C or public CDN A delegates traffic there.  Even though routing should be smart enough so that public CDN A routes OC CDN C's traffic directly instead of through OC CDN B (unless there is a peering failure :-) ), this could still result in triggers being propagated from C to A to B, and back to C again.  So, the safe thing to do would be for CDN B to not fail the trigger, but to report successful trigger processing and not propagate it directly to C.

Reply:
I was thinking that the safe thing to do would be to propagate it directly to C, and if C is the one originating the trigger, it would respond to it accordingly
In other words, have CDNs involved in a chain, detect loops for themselves rather than try to detect it (and fail a trigger) on behalf of other CDNs in the chain

Reply:
rewrote as 
   <t>             
   dCDN that receives a trigger creation request that contains "cdn-path" attribute SHOULD check it for its own CDN PID.
   If dCDN's PID is already present and dCDN is not the CDN initiating the trigger, this condition likely indicates a loop,
   In such case, dCDN MUST reject the trigger, resulting in trigger being rejected back to its originating uCDN.
   If dCDN receives a trigger that it itself originated, dCDN MAY process the trigger as required.
   </t>
   <t>     
   dCDN that cascades trigger requests to additional dCDNs (so-called "transit CDN"), SHOULD NOT reject
   triggers that have CDN PID of its downstream CDNs in their CDN path, allowing each CDN to do their own loop detection.
   </t>